validator/rules/known_root_type.go

```diff
@@ -0,0 +1,35 @@
+package validator
+
+import (
+	"fmt"
+
+	"github.com/vektah/gqlparser/v2/ast"
+	. "github.com/vektah/gqlparser/v2/validator"
+)
+
+func init() {
+	AddRule("KnownRootType", func(observers *Events, addError AddErrFunc) {
+		// A query's root must be a valid type.  Surprisingly, this isn't
+		// checked anywhere else!
+		observers.OnOperation(func(walker *Walker, operation *ast.OperationDefinition) {
+			var def *ast.Definition
+			switch operation.Operation {
+			case ast.Query, "":
+				def = walker.Schema.Query
+			case ast.Mutation:
+				def = walker.Schema.Mutation
+			case ast.Subscription:
+				def = walker.Schema.Subscription
+			default:
+				// This shouldn't even parse; if it did we probably need to
+				// update this switch block to add the new operation type.
+				panic(fmt.Sprintf(`got unknown operation type "%s"`, operation.Operation))
+			}
+			if def == nil {
```

Comment:
Am I wrong to think that def cannot be nil here due to an unsupported operation type, given that it would have panicked above in the default switch? 
I'm not sure under what circumstances that `walker.Schema.{Subscription|Mutation|Query}` would be nil, but I haven't played with this area in a while.

Reply:
This gets hit in the case we are interested in -- you wrote `mutation { ... }` but the schema has no type `Mutation`.  The panic is only if you wrote `bogus { ... }`, i.e. a totally invalid operation type, and that we should have rejected earlier (which is why it's a panic).

Reply:
Ah, right. the error message case is when it is invalid in the case of this particular schema.